### ISSUE TITLE
API example

### ISSUE THREAD

**Issue by jakobrosenberg** [PepsRyuu/nollup#96]
Hi. Do you have a working example of the API? I haven't been able to get it working. I Keep getting `SyntaxError: 'import' and 'export' may only appear at the top level.` The error seems to affect all files in the project.

**Comment by PepsRyuu** [comment]
Hi @jakobrosenberg 

The [API](https://github.com/PepsRyuu/nollup/blob/master/API.md) docs provides a barebones example at the very top. Is this the one you're referring to?

Although I suspect from the error that it's a plugin configuration issue. Can you provide the ```rollup.config.js``` file?

**Comment by jakobrosenberg** [comment]
Thanks for the the quick reply! Yes, that's the one.

We're working on a new framework, Roxi. It's going to abstract the wiring from [Routify](https://routify.dev)'s starter template. The config has been tested and works with the CLI, so I figured it would be a good starting point.

[rollup.config.js](https://github.com/sveltech/routify-starter/blob/beta/rollup.config.js) (taken from Routify's starter template)

<details>
<summary>
Implementation in Roxi
</summary>


```javascript
const rollup = require('rollup')
const nollup = require('nollup')
const path = require('path');
const loadConfigFile = require('rollup/dist/loadConfigFile');

async function startNollup() {
  process.env.NOLLUP = true
  process.env.ROLLUP_WATCH = true
  // load the config file next to the current script;
  // the provided config object has the same effect as passing "--format es"
  // on the command line and will override the format of all outputs
  loadConfigFile(path.resolve(__dirname, 'rollup.config.js'), { format: 'es' })
    .then(async ({ options, warnings }) => {



      // "warnings" wraps the default `onwarn` handler passed by the CLI.
      // This prints all warnings up to this point:
      console.log(`We currently have ${warnings.count} warnings`);

      // This prints all deferred warnings
      warnings.flush();

      // options is an "inputOptions" object with an additional "output"
      // property that contains an array of "outputOptions".
      // The following will generate all outputs and write them to disk the same
      // way the CLI does it:
      options.forEach(async options => {
        const bundle = await nollup(options)
        bundle.generate(options.output);
      })
    })
}
  module.exports = { startNollup }

```
</details>

I realize it's a relatively complex, which is why I was hoping you had an example I could clone ^^

**Comment by PepsRyuu** [comment]
Great! Is there a step by step process I can use to replicate the error? Clone the starter, run the script?

**Comment by jakobrosenberg** [comment]
If you npm link the template to the framework and run `npm run dev`, you should see what I see. At the moment that's not a lot, since the errors are gone with commonjs 13, but there's still nothing on port 8080.

Framework: https://github.com/sveltech/roxi
Template: https://github.com/sveltech/roxi-template

**Comment by PepsRyuu** [comment]
Ah there might be a bit of a misunderstanding on how the API works here. The low-level API will only generate a bundle, it doesn't run any web server at all. It's basically identical to Rollup, where Rollup would typically expect you to either wrap Rollup in a web-server, or use their ```write``` or ```watch``` API which writes to disk. Nollup doesn't provide a ```write()/watch()``` function as it would mean writing to disk frequently which isn't great for the lifespan of the disk. So instead, it provides a middleware which serve the bundles from memory. 

There's three levels of APIs for Nollup:

* **API** - Generates a bundle, basically identical to the Rollup API. Also on each bundle provides changed modules for the purpose of implementing HMR. There's no web server here, it just provides compiled code.

* **Dev Middleware** - Wraps Nollup into an Express middleware. It compiles the bundle and exposes those compiled files. Watches the file system for changes, and provides a HMR implementation. You can see information about it here: https://github.com/PepsRyuu/nollup/blob/master/API.md#dev-middleware.

* **CLI** - Wraps dev middleware into an easy to use CLI (```nollup -c```), but also adds other typical web server features such as static file serving, index fallback and proxying.

It sounds like for your use case you might want to consider using the dev middleware.

**Comment by PepsRyuu** [comment]
Just looking through some of the code, you could do the same technique used for Snowpack:

```
function startNollup() {
  process.argv.push('-c', '_roxi/rollup.config.js')
  process.argv.push('--port', '5000')
  process.argv.push('--hot');
  process.argv.push('--content-base', 'public')
  require('nollup/lib/cli');
}

module.exports = { startNollup }
```

That ironically enough triggered the ```import``` top-level issue you were having. Switching to ```rollup-plugin-commonjs-alternate``` allows it to compile (it's a CommonJS plugin I wrote a while back before the official plugin did its rewrite, so there's a bug with that one). Still shows a blank screen, but accessing ```http://localhost:5000/build/bundle.js``` works just fine then. :)

**Comment by jakobrosenberg** [comment]
Thanks, @PepsRyuu. I had completely forgotten about my CLI hack for Snowpack. 

Have you considered abstracting https://github.com/PepsRyuu/nollup/blob/master/lib/cli.js#L92-L136 into `watch.js` and call it with `require('./watch')(options)`

It would provide a clean and predictable watch API, since it would share documentation, usage and expectations with the CLI.

I can create a PR if you're interested.

**Comment by PepsRyuu** [comment]
I'd totally be up for that. I wouldn't call it ```watch``` though, probably just ```dev-server```. This would make it very similar to how Webpack structures ```webpack-dev-server``` and ```webpack-dev-middleware```.

So there would be 4 ways of using Nollup:

* CLI - Simple CLI, uses dev server and passes the options to it.
* Dev Server - Reads nolluprc, converts Rollup config to ESM, runs Express server on port.
* Dev Middleware - Same as described earlier.
* API - Same as described earlier.

```
require('nollup/lib/dev-server')({
    config: 'rollup.config.js',
    port: 9001,
    before: () => {},
    after: () => {},
    contentBase: '',
    // all other cli options
});
```

**Comment by jakobrosenberg** [comment]
Great! I will send a PR shortly.

I agree with the naming. `watch` is a misnomer in this case.

**Comment by PepsRyuu** [comment]
Released in ```0.12.0```
